crowbar_framework/lib/crowbar/chef/upload.rb

```diff
@@ -0,0 +1,284 @@
+#
+# Copyright 2011-2013, Dell
+# Copyright 2011-2016, Chef Software Inc.
+# Copyright 2013-2016, SUSE Linux GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/cookbook_uploader"
+require "chef/knife/core/object_loader"
+
+module Crowbar
+  class Chef
+    class Upload
+      class << self
+        def all
+          cookbooks
+          data_bags
+          roles
+        end
+
+        def cookbooks
+          version_constraints_to_update = Hash.new
+          cookbooks_for_upload = Array.new
+
+          cookbook_loader.each do |cookbook_name, cookbook|
+            cookbooks_for_upload.push(cookbook)
+            version_constraints_to_update[cookbook_name] = cookbook.version
+          end
+
+          logger.info("Checking existing cookbooks...")
+          cookbooks_for_upload.reject! do |cookbook|
+            next unless api_coobbooks.include?(cookbook.name.to_s)
+            cookbook_exists?(cookbook)
```

Comment:
thanks for outsourcing this. only `cookbook_exist?` is more conventional, can you change it?